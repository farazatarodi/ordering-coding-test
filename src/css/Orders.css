.orders-container {
  width: 80%;
  height: fit-content;
}

.orders-content {
  margin: 3rem 0;
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  column-gap: 2%;
  align-self: center;
  border: solid 2px var(--color-neutral);
  box-shadow: var(--shadow);
  border-radius: var(--radius);
}

.orders-header {
  width: 100%;
  display: grid;
  grid-template-columns: 10% auto 20% 20%;
  padding: 2rem;
  background-color: var(--color-neutral-light);
  border-radius: var(--radius) var(--radius) 0 0;
}
