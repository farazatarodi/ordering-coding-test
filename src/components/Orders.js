// main Orders page component

// Redux imports
import { useSelector } from 'react-redux';

// Router imports
import { Switch, Route, useRouteMatch } from 'react-router-dom';

// component imports
import OrderCard from './OrderCard';
import Order from './Order';
import EmptyCard from './EmptyCard';

// css imports
import '../css/Orders.css';

const Orders = () => {
  // gert all orders from store
  const orders = useSelector((state) => state.ordersReducer);

  // matching current address (/orders)
  const match = useRouteMatch();

  return (
    <div className="orders-container">
      <h1 className="heading heading-1">Orders</h1>
      {/* secondary switch affects the orders page */}
      <Switch>
        {
          // map orders to order routes, address is order ID
          orders &&
            orders.map((order) => (
              <Route key={order.id} path={`${match.path}/${order.id}`}>
                <Order id={order.id} />
              </Route>
            ))
        }

        {/* Orders page base route */}
        <Route path={match.path}>
          <div className="orders-content">
            <div className="orders-header heading heading-2">
              <strong>ID</strong>
              <strong>Customer</strong>
              <strong>Total</strong>
              <strong>Details</strong>
            </div>
            {
              // map orders to order card components
              orders && orders.length !== 0 ? (
                orders.map((order) => (
                  <OrderCard
                    key={order.id}
                    total={order.total}
                    customerId={order['customer-id']}
                    orderId={order.id}
                  />
                ))
              ) : (
                <EmptyCard />
              )
            }
          </div>
        </Route>
      </Switch>
    </div>
  );
};

export default Orders;
