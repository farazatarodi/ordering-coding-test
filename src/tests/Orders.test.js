import { render, screen, fireEvent } from '@testing-library/react';

import Orders from '../components/Orders';

import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';

import { BrowserRouter as Router, Route, Redirect } from 'react-router-dom';

import renderer from 'react-test-renderer';

test('we see the empty table in orders page', () => {
  const initialState = { ordersReducer: { orders: [] } };
  const mockStore = configureStore();
  const store = mockStore(initialState);

  render(
    <Provider store={store}>
      <Router>
        <Orders />
      </Router>
    </Provider>
  );

  expect(screen.getByText(/nothing/i)).toBeTruthy();
});

test('order card shows all info correctly', () => {
  const initialState = {
    ordersReducer: {
      orders: [
        {
          id: 'testOrderId',
          'customer-id': 'testCustomerID',
          items: [],
          total: 'testTotal',
        },
      ],
    },
    customersReducer: {
      customers: [
        {
          id: 'testCustomerID',
          name: 'testName',
          since: 'testSince',
          revenue: 'testRevenue',
        },
      ],
    },
  };

  const mockStore = configureStore();
  const store = mockStore(initialState);

  render(
    <Provider store={store}>
      <Router>
        <Orders />
      </Router>
    </Provider>
  );

  expect(screen.getByText(/testorderid/i)).toBeTruthy();
  expect(screen.getByText(/testname/i)).toBeTruthy();
  expect(screen.getByText(/testtotal/i)).toBeTruthy();
});

test('"details" and "back" buttons work', () => {
  const initialState = {
    ordersReducer: {
      orders: [
        {
          id: 'testOrderId',
          'customer-id': 'testCustomerID',
          items: [],
          total: 'testTotal',
        },
      ],
    },
    customersReducer: {
      customers: [
        {
          id: 'testCustomerID',
          name: 'testName',
          since: 'testSince',
          revenue: 'testRevenue',
        },
      ],
    },
    productsReducer: {
      products: [],
    },
  };

  const mockStore = configureStore();
  const store = mockStore(initialState);

  render(
    <Provider store={store}>
      <Router>
        <Route path="/orders">
          <Orders />
        </Route>
        <Redirect to="/orders" />
      </Router>
    </Provider>
  );

  fireEvent.click(screen.getByRole('link', { name: /details/i }));
  expect(screen.getByRole('link', { name: /back/i })).toBeTruthy();
  expect(screen.queryByRole('link', { name: /details/i })).toBeNull();

  fireEvent.click(screen.getByRole('link', { name: /back/i }));
  expect(screen.getByRole('link', { name: /details/i })).toBeTruthy();
  expect(screen.queryByRole('link', { name: /back/i })).toBeNull();
});

test('we see the loading card while fetching data', () => {
  const initialState = {
    ordersReducer: {
      orders: [],
      loading: true,
      error: null,
    },
  };

  const mockStore = configureStore();
  const store = mockStore(initialState);

  render(
    <Provider store={store}>
      <Router>
        <Orders />
      </Router>
    </Provider>
  );

  expect(screen.getByText(/Loading/i)).toBeTruthy();
});

test('we see the error data if fetch fails', () => {
  const initialState = {
    ordersReducer: {
      orders: [],
      loading: false,
      error: 'testError',
    },
  };

  const mockStore = configureStore();
  const store = mockStore(initialState);

  render(
    <Provider store={store}>
      <Router>
        <Orders />
      </Router>
    </Provider>
  );

  expect(screen.getByText(/Error Fetching Data: testError/i)).toBeTruthy();
});

test('renders correclty', () => {
  const initialState = {
    customersReducer: { customers: [] },
    ordersReducer: { orders: [] },
    productsReducer: { products: [] },
  };
  const mockStore = configureStore();
  const store = mockStore(initialState);
  const tree = renderer
    .create(
      <Provider store={store}>
        <Router>
          <Orders />
        </Router>
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
