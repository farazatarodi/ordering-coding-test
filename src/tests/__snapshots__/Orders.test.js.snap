// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correclty 1`] = `
<div
  className="orders-container"
>
  <h1
    className="heading heading-1"
  >
    Orders
  </h1>
  <div
    className="orders-content"
  >
    <div
      className="orders-header heading heading-2"
    >
      <strong>
        ID
      </strong>
      <strong>
        Customer
      </strong>
      <strong>
        Total
      </strong>
      <strong>
        Details
      </strong>
    </div>
    <div
      className="card text text-display notification-card"
    >
      Nothing to show
    </div>
  </div>
</div>
`;
